### PR TITLE
WD-2175 Career progression

### PR DESCRIPTION
**Pull request by carkod**
## Done

Build career/progression page
## QA

- View https://canonical-com-823.demos.haus/careers/progression and compare with [design](https://www.figma.com/file/UsovUAf4zELQgGd9bxhiVe/Careers?node-id=1496-11228&t=ZOwzdx88YOIY8Rnm-0)
- Check everything looks good on mobile
- Check the other careers search pages work (this PR has refactored search) e.g. https://canonical-com-823.demos.haus/careers/all and https://canonical-com-823.demos.haus/careers. Do a quick search
- Run through the following [QA steps](https://canonical-web-and-design.github.io/practices/workflow/qa-steps.html)
- [List additional steps to QA the new features or prove the bug has been resolved]

## Fun fact/Behind-the-scenes

[About 6 codepens](https://codepen.io/carkod/pens/public?cursor=ZD0wJm89MSZwPTEmdj02NzE0ODcyNQ==) were created before putting together this page.

## Issue / Card

Fixes 

## Screenshots

[if relevant, include a screenshot]

### CODE REVIEW

File: static/js/careers-progression.js
Comment (anthonydillon) [inline static/js/careers-progression.js:218]: Function names should be camelcase

**Review by anthonydillon — CHANGES_REQUESTED** [review]
Wow @carkod this looks great. I don't think you are too far off. I left some comments in the code and there are some improvements to the small screen view.

**Comment by carkod** [comment]
> Hi @carkod Thank you for building this so fast. It looks cool! I do have feedback smile
> 
> Screen Shot 2023-03-23 at 1 13 30 PM
> 
> * the line should be inside the column with 22px padding-right.
> * the line opacity is a 0.25
> * there's a m,issing line on the leftmost content.
> 
> Screen Shot 2023-03-23 at 1 14 08 PM
> 
> * Is it possible that the photos genuinely shuffle back and smoothly go into a loop with the same content?
> * The text isn't changing as you click through
> 
> Screen Shot 2023-03-23 at 1 14 55 PM
> 
> - add break
> ```
> <div><h#> To deliver the world's best open source <br> experience, from platform to application<h#><div>
> ```
> 
> * remove break on small screen sizes.
> 
> Screen Shot 2023-03-23 at 1 15 05 PM
> 
> - the background color of the card has some transparency just make it a solid #fff 
> Thank you! I'll make a mobile version view of the whole thing. So it translates better in smaller screen sizes. I found issues with the mobile view that can be explained better visually. Again thank you for your hard work on this, Carlos. I genuinely appreciate it.

Addressed all comments!

File: static/sass/_pattern_careers-progression.scss
Comment (carkod) [reply static/sass/_pattern_careers-progression.scss:104]: I left these comments on purpose in case design wants to change some of the animations. Once I get design review +1 I'll remove them.

File: static/sass/_pattern_careers-progression.scss
Comment (carkod) [reply static/sass/_pattern_careers-progression.scss:30]: Removing them causes the images not to align and design has very specific spacing that is not even and there is a border on the left, all those styles are covered by this class.

File: static/sass/_pattern_careers-progression.scss
Comment (carkod) [reply static/sass/_pattern_careers-progression.scss:75]: This is overriding the current Vanilla styles which cause the element to align in the center and some other additional padding, it's not a utility to add styling it's correction. Vanilla base styles add styles to the `video` element by defeault

File: static/sass/_pattern_careers-progression.scss
Comment (carkod) [reply static/sass/_pattern_careers-progression.scss:4]: I can use the function, but the color is obviously not the same.

File: static/sass/_pattern_careers-progression.scss
Comment (carkod) [reply static/sass/_pattern_careers-progression.scss:56]: That is not intended for reuse. This hero section has anitmation that other pages don't have, and has a lot more padding and spacing than other careers pages. I could append the `careers-progression` prefix, but the class name was getting very long...

**Comment by marinacastejon** [comment]
Hi @carkod I found a few things:
- The images are not the same height
image
- The chevrons are not center-aligned in the 2 sections where they appear
image
- The video is not accessible the way it is, I'm leaving a comment for @camillet16  in Figma proposing accessible alternatives. 
Thanks, 
Marina

File: static/js/careers-progression.js
Comment (anthonydillon) [reply static/js/careers-progression.js:17]: Did you try this @carkod?

File: static/js/careers-progression.js
Comment (anthonydillon) [reply static/js/careers-progression.js:21]: Did you try this @carkod?

File: static/js/careers-progression.js
Comment (anthonydillon) [inline static/js/careers-progression.js:274]:
```suggestion
   * Layout to hug the right edge of the page
```

File: static/js/careers-progression.js
Comment (anthonydillon) [inline static/js/careers-progression.js:272]: I believe this would be more performant as CSS. Example: https://codepen.io/anthonydillon/pen/JjaVYJR?editors=1100

File: static/sass/_pattern_careers-progression.scss
Comment (anthonydillon) [inline static/sass/_pattern_careers-progression.scss:123]: The linear easing here feels very robotic.

File: static/js/careers-progression.js
Comment (anthonydillon) [reply static/js/careers-progression.js:21]: This is setting an undefined class

**Review by anthonydillon — CHANGES_REQUESTED** [review]
Few more comments there @carkod. Let me know if you want to discuss any.

File: static/js/careers-progression.js
Comment (carkod) [reply static/js/careers-progression.js:17]: Yes, because this needs to add an observer to every element, that needs to animate, it doesn't seem to work.

File: static/js/careers-progression.js
Comment (carkod) [reply static/js/careers-progression.js]: Fixed

File: static/sass/_pattern_careers-progression.scss
Comment (carkod) [reply static/sass/_pattern_careers-progression.scss]: Added more custom `cubic-benzier` but let me know if you want something more specific https://cubic-bezier.com/

**Comment by carkod** [comment]
@marinacastejon @camillet16 can I get another review? Thanks

**Comment by carkod** [comment]
> @carkod @camillet16 After reading Carlos' comments in figma, I think it might be worth re-prioritising things and postponing the video. If to make it accessible it's going to be a lot of work having the video, then I suggest launching this only with an image and creating a separate issue just to work on the accessible video. I think there are other things with a higher priority in this project, such as the meganav. What do you think? FYI @anthonydillon

Well, adding controls is ok. It's already done, if you check the demo. But maybe video accessibility is more than just adding controls?

**Comment by anthonydillon** [comment]
@carkod The video looks good to me. The only a11y guideline not hit is the ability to hide the video but I think the pause is good for now.

File: static/js/careers-progression.js
Comment (carkod) [reply static/js/careers-progression.js:23]: Why not? This is what's being used on the homepage?

File: static/sass/_pattern_careers-progression.scss
Comment (carkod) [reply static/sass/_pattern_careers-progression.scss]: Removed

File: static/sass/_pattern_careers-progression.scss
Comment (carkod) [reply static/sass/_pattern_careers-progression.scss]: Thanks. Moved to the title class

File: static/js/careers-progression.js
Comment (anthonydillon) [reply static/js/careers-progression.js:23]: s/wouldn't/would

File: templates/careers/progression.html
Comment (anthonydillon) [inline templates/careers/progression.html:35]:
```suggestion
      <p>Not just how to be better engineers but how to be better speakers, how to be better designers, better organisers, better partners, and better vendors.</p>
```

File: templates/careers/progression.html
Comment (anthonydillon) [inline templates/careers/progression.html:92]: This online style can be more elegantly handled in CSS `last-of-type`

File: templates/careers/progression.html
Comment (anthonydillon) [inline templates/careers/progression.html:189]: Does this sentence describe the image to someone that can't see it?

File: templates/careers/progression.html
Comment (anthonydillon) [inline templates/careers/progression.html:208]: Does this sentence describe the image to someone that can't see it?

File: templates/careers/progression.html
Comment (anthonydillon) [inline templates/careers/progression.html:218]:
```suggestion
      <h3>Ubuntu Core is set to redefine industrial computing with new edge AI platform NVIDIA IGX</h3>
```

File: templates/careers/progression.html
Comment (anthonydillon) [inline templates/careers/progression.html:247]: Could you merge this section with the one below somehow as they are connected content?

File: templates/careers/progression.html
Comment (anthonydillon) [inline templates/careers/progression.html:253]: It's a shame to not have citations for these quotes. It would make them more personal and real

File: templates/careers/progression.html
Comment (anthonydillon) [inline templates/careers/progression.html:277]: Never seen `u-fixed-width` and `p-strip` used together. Was this by design?

File: templates/careers/progression.html
Comment (anthonydillon) [inline templates/careers/progression.html:302]: And again but doubled up?

File: webapp/app.py
Comment (anthonydillon) [inline webapp/app.py:225]: Is this logic now duplicated? If so, we should centralise it with a reuseable util.

**Review by anthonydillon — CHANGES_REQUESTED** [review]
Really good work here @carkod. Some general comments in the code.

File: templates/careers/progression.html
Comment (anthonydillon) [inline templates/careers/progression.html:217]:
```suggestion
      <small>20 September 2022</small>
```
This is the future

**Review by anthonydillon — COMMENTED** [review]
The line height on the hero title seems too large on small screens.


There is a little x-axis scrolling in the phone.

File: templates/careers/progression.html
Comment (carkod) [reply templates/careers/progression.html:189]: Well, I can remove it, but the a11y validator complains of missing `alt`

**Comment by carkod** [comment]
> @carkod The video looks good to me. The only a11y guideline not hit is the ability to hide the video but I think the pause is good for now.

The video hides with screen readers via `aria-hidden`

File: webapp/app.py
Comment (carkod) [reply webapp/app.py:225]: Ok I think I found a way to achieve both.

File: templates/careers/progression.html
Comment (carkod) [reply templates/careers/progression.html:92]: So I think the trade-off is that it will become more messy in the stylesheet. It's not exactly `last-child` it's actually last-child of a different parent i.e. a "cousin". So to use a selector for the cousin means I will need to create a parent selector (new parent selector? Because I'm using Vanilla styles mostly here, no custom parent class) that goes down to the child, just so I can add one line of CSS.

**Comment by carkod** [comment]
> The line height on the hero title seems too large on small screens.
> 
> There is a little x-axis scrolling in the phone.

Fixed

File: webapp/app.py
Comment (carkod) [inline webapp/app.py:80]: This is just moving the code up, so that all `_` helper functions are in one place to be referenced.

**Comment by carkod** [comment]
> * https://canonical-com-823.demos.haus/careers/progression

Fixed

File: static/js/careers-progression.js
Comment (anthonydillon) [inline static/js/careers-progression.js:14]: videoElement here will be an array. You are likely to hit an error.

File: static/sass/_pattern_careers-progression.scss
Comment (anthonydillon) [inline static/sass/_pattern_careers-progression.scss:6]: This will affect all body elements on this site. Is that expected? If so, I would suggest putting it somewhere central like styles.scss

File: static/sass/_pattern_careers-progression.scss
Comment (anthonydillon) [inline static/sass/_pattern_careers-progression.scss:113]: Won't this animate all the separators across the site?

File: tests/test_routes.py
Comment (anthonydillon) [inline tests/test_routes.py:78]: 👍

**Review by anthonydillon — APPROVED** [review]
LGTM Carlos, few small comments but good to go once addressed 👍

File: static/sass/_pattern_careers-progression.scss
Comment (carkod) [reply static/sass/_pattern_careers-progression.scss:113]: I couldn't achieve the same effect with the Vanilla styles, so I renamed it as a custom class

File: static/sass/_pattern_careers-progression.scss
Comment (carkod) [reply static/sass/_pattern_careers-progression.scss]: This has been removed now.

**Pull request merged**
